### ISSUE TITLE
Adding new hooks

### ISSUE THREAD

**Issue by aress31** [OpenZeppelin/openzeppelin-contracts#2674]
The `beforeTokenTransfer` hook is super useful and make it easy to extend `ERC20` core functionalities. However, using this same hook in many different functions breaks the code flexibility - as different functions may need different logic - therefore, what I suggest is that instead of exporting only `beforeTokenTransfer`, you could expose:
- `beforeMint`: for `_mint`
- `beforeBurn`: for `_burn`
- `beforeTokenTransfer`: for `_transfer`

Also having a way to update  `_balances` directly from children contracts without having to redefine all relevant functions:
- `balanceOf`
- `_mint`
- `_burn`
- `_transfer`

Would help avoiding code repetition and save in gas fees for deployment.

**Comment by Amxx** [comment]
Hello @aress31,

The common thing with mint, burn and transfer, is that they all affect the balances of the wallet, either by adding of removing tokens. This is what dev most often want to catch using the `_beforeTransfer` hook.

Having three differents hook will likely cause code duplication, which is not a good thing. Also, from within a `_beforeTransfer` hook, you can check if `from` or `to` is `address(0)`. 

For gas saving purposes (avoid the check in `_beforeTransfer`) you can also, as you stated, override `_mint`. This is what we do in the [ERC20Capped](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC20/extensions/ERC20Capped.sol) contract.

In conclusion, I do believe that our current hook system is powerful enough for most cases. I have never found a case where I needed more granular hooks, and the ability to build ERC20 extensions is giving me confidence.

I'd love to know more about your usecases, and the gas savings you would get from such hooks.

**Comment by aress31** [comment]
@Amxx then similarly to the `allowance` an `increaseBalance` and `decreaseBalance` functions should be expose so that in my own contract, I do not have to redefine the `_balances` array and rewrite the relevant functions (`mint`, `transfer` and `burn`) whenever wanting to do something with balances in my custom contract!

**Comment by frangio** [comment]
@aress31 `increaseBalance` and `decreaseBalance` are already available, in the form of `_mint` and `_burn` respectively. These, together with `_transfer`, are the only way to update the balances mapping. The reason it's done this way is that it's important to also emit a `Transfer` event whenever the mapping is changed, which is done by all three functions.

As @Amxx said, you should be able to do everything you want with `_beforeTokenTransfer`, and in some cases you may want to override `_mint` or `_burn` directly.

**Comment by aress31** [comment]
True haven't thought of mint and burn as a way to alter the balance in the
same way than the increase/decrease allowances. Thanks!
